File: Introduction-to-Algorithms/第二章（非线性排序）/2.1插入排序/python/main.py
# 算法导论-第二章-2.1
# 插入排序
# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# INSERTION-SORT（A）
# for j <- 2 to length[A]
#     do key <- A[j]
#         Insert A[j] into the sorted sequence A[1...j-1].
#         i <- j-1
#         while i>0 and A[i]>key
#             do A[i+1] <- A[i]
#                 i <- i-1
#         A[i+1] <- key
# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
# 2014. 09. 08
# lichao

import random

# 产生随机数组
# N为数组长度
def gen_rand_n(N):
	A = []
	for i in range(1, N+1):
		A.append(random.randint(1,N*N))
	return A

# 插入排序
# 从A[0]开始
def insertion_sort(A, N):
	for j in range(0, N):
		key = A[j];
		i = j-1;

		while (i>=0) & (A[i]>key):
			A[i+1] = A[i]
			i = i-1

		A[i+1] = key

	return A

# 主函数
N = 10
A = gen_rand_n(N)

print (A)

A = insertion_sort(A, N)

print (A)

File: Introduction-to-Algorithms/第二章（非线性排序）/2.1插入排序/c++/main.cpp

// 算法导论-第二章-2.1
// 插入排序
// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
// INSERTION-SORT（A）
// for j <- 2 to length[A]
//     do key <- A[j]
//         Insert A[j] into the sorted sequence A[1...j-1].
//         i <- j-1
//         while i>0 and A[i]>key
//             do A[i+1] <- A[i]
//                 i <- i-1
//         A[i+1] <- key
// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
// 2014. 09. 08
// lichao

#include <iostream>
#include <stdlib.h>

using namespace std;

// 产生随机数组
template<size_t SIZE>
void gen_rand_n(int (&A)[SIZE])
{
	srand(5);
	for (int i = 0; i < SIZE; ++i)
	{
		A[i] = rand();
	}
}

// 插入排序
template<size_t SIZE>
void insertion_sort(int (&A)[SIZE])
{
	int key = 0;
	int j = 0;

	for (int i = 1; i < SIZE; ++i)
	{
		key = A[i];
		j = i - 1;
		while((j >= 0) && (A[j] > key))
		{
		    A[j+1] = A[j];
		    j = j-1;
		}
		A[j+1] = key;
	}
}

int main()
{	
	const int N(100); // 定义数组长度

	int A[N] = {0}; // 定义数组
	gen_rand_n(A); // 产生随机数组

	// for (int i = 0; i < N; ++i)
	// {
	// 	cout << A[i] << endl;
	// }

	insertion_sort(A);

	for (int i = 0; i < N; ++i)
	{
		cout << A[i] << endl;
	}

	return 0;
}


